Userland/Services/WindowServer/Overlays.cpp

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) 2021, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "Overlays.h"
+#include "Compositor.h"
+#include "WindowManager.h"
+
+namespace WindowServer {
+
+Overlay::~Overlay()
+{
+    Compositor::the().remove_overlay(*this);
+}
+
+void Overlay::invalidate()
+{
+    // m_current_rect should only get updated by recompute_overlay_rects()
+    if (!m_current_rect.is_empty()) {
+        Compositor::the().invalidate_screen(m_current_rect);
+        m_current_rect = {};
+    }
+}
+
+void Overlay::set_rect(Gfx::IntRect const& rect)
+{
+    if (m_rect == rect)
+        return;
+    m_rect = rect;
+    invalidate();
+    rect_changed();
+    if (has_been_added())
+        Compositor::the().overlay_rects_changed();
+}
+
+BitmapOverlay::BitmapOverlay()
+{
+    clear_bitmaps();
+}
+
+void BitmapOverlay::rect_changed()
+{
+    clear_bitmaps();
+}
+
+void BitmapOverlay::clear_bitmaps()
+{
+    m_bitmaps = MultiScaleBitmaps::create_empty();
+}
+
+void BitmapOverlay::render(Gfx::Painter& painter, Screen const& screen)
+{
+    auto scale_factor = screen.scale_factor();
+    auto* bitmap = m_bitmaps->find_bitmap(scale_factor);
+    if (!bitmap) {
+        auto new_bitmap = create_bitmap(scale_factor);
+        if (!new_bitmap)
+            return;
+        bitmap = new_bitmap.ptr();
+        m_bitmaps->add_bitmap(scale_factor, new_bitmap.release_nonnull());
+    }
+
+    painter.blit({}, *bitmap, bitmap->rect());
+}
+
+ScreenNumberOverlay::ScreenNumberOverlay(Screen& screen)
+    : m_screen(screen)
+{
+    Gfx::IntRect rect {
+        default_offset,
+        default_offset,
+        default_size,
+        default_size
+    };
+    if (auto* overlay_bitmaps = WindowManager::the().screen_number_bitmap())
+        rect.set_size(overlay_bitmaps->default_bitmap().size());
+    rect.translate_by(screen.rect().location());
+    set_rect(rect);
+}
+
+RefPtr<Gfx::Bitmap> ScreenNumberOverlay::create_bitmap(int scale_factor)
+{
+    auto* font = Compositor::the().screen_number_font();
+    if (!font)
+        return nullptr;
+    auto bitmap = Gfx::Bitmap::create(Gfx::BitmapFormat::BGRA8888, rect().size(), scale_factor);
+    if (bitmap) {
+        Gfx::Painter painter(*bitmap);
+        if (auto* overlay_bitmaps = WindowManager::the().screen_number_bitmap()) {
+            auto& overlay_bitmap = overlay_bitmaps->bitmap(scale_factor);
+            painter.blit({}, overlay_bitmap, overlay_bitmap.rect());
+        } else {
+            painter.fill_rect(bitmap->rect(), Color(Color::Black).with_alpha(0xcc));
```

Comment:
```suggestion
            painter.fill_rect(bitmap->rect(), Color(Color::Black).with_alpha(0xcc));
            painter.fill_rect_with_rounded_corners(bitmap->rect(), Color(Color::Black).with_alpha(0xcc), 10);
```

![image](https://user-images.githubusercontent.com/388571/122982769-92219c80-d39b-11eb-9ba1-c3c3565d641c.png)

Reply:
Try this with 2x scaling and you'll know why I didn't do this (yet) ;-)